### DISCUSSION TITLE
Returning ORM instances from arbitrary query

### DISCUSSION THREAD

**Discussion by al-dpopowich** [sqlalchemy/sqlalchemy#7479] [Usage Questions]
I asked this question over at [stackoverflow](https://stackoverflow.com/questions/70326764/using-arbitrary-sqlalchemy-select-results-e-g-from-a-cte-to-create-orm-insta), but I'm receiving no love there, so asking here: 

There are two ways (that I know of) to return model instances with sqla:

1. Use `session.get(MyModel, pk)`.
2. Use `select(MyModel)` to generate a query for later use with `session.execute()`.

But what about an arbitrary query that returns rows of a mapped table?  E.g., maybe I have a complex query that returns rows of a table via a function call: `select(func.fetch_rows_of_mymodel(arg1, arg2))`.  My stackoverflow example is a complex series of CTE that returns a newly created row in which case I know the row returned is "clean" and would like an instance in the current session that is in state persistent.  And note: `session.merge()` is not the answer I'm looking for as that issues another round trip to the db.

Is there a way to "annotate" an arbitrary query to return instances of a model, persistent in the session?  I'm hoping there's something obvious I've overlooked!

Thanks!

**Comment by zzzeek** [comment]
yes, for this use case you use the [aliased](https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.aliased) construct.   for examples of how to convert individual columns into entities when you fetch results including from CTEs, see https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#orm-entity-subqueries-ctes

**Reply by al-dpopowich** [reply]
Thanks for the tip on using `aliased()`, but I'm still having trouble with this.  The example you link to does not exactly line up with my use-case. I have a chain of CTE and then want to return a row from one of the CTE. Here is the SQL I'm trying to model with python:

With this table:

```sql
create table company (
  id serial primary key,
  name varchar not null
);
```

This is the query:

```sql
with new_company as (
     insert into company (name)
          values ('Acme, Inc.')
       returning *
),

other_related as (
   -- herein we join to `new_company` and create default related rows
   -- in other tables.  Here we use, effectively, a no-op as what it
   -- does is not germane to the issue.
   select id from new_company
)

-- Having created the related rows, we return the row we inserted into
-- table `company`.
select * from new_company;
```

With sqlalchemy I can do this:

```python
import sqlalchemy as sa
import sqlalchemy.orm


Base = sa.orm.declarative_base()

_Company = sa.Table(
   'company',
   Base.metadata,
   sa.Column('id', sa.Integer, primary_key=True),
   sa.Column('name', sa.Text, nullable=False),
)

class Company(Base):

   __table__ = _Company


new_company = (
   sa.insert(Company)
   .values(name='Acme, Inc.')
   .returning(Company)
   .cte(name='new_company')
)
other_related = (
   sa.select(sa.text('new_company.id'))
   .select_from(new_company)
   .cte('other_related')
)
fetch_company = (
   sa.select(sa.text('* from new_company'))
   .add_cte(other_related)
)
print(fetch_company)
```

And executing the above produces:

```
WITH new_company AS 
(INSERT INTO company (name) VALUES (:param_1) RETURNING company.id, company.name), 
other_related AS 
(SELECT new_company.id 
FROM new_company)
 SELECT * from new_company
```

And as stated in my original question, I can return rows, but not instances of `Company` that are in persisted state.

I've tried using `aliased()`, but I haven't figured out (yet) how to generate a query that doesn't produce an error ("Query contains no columns with which to SELECT from") or doesn't end up doing a cross join.

**Reply by zzzeek** [reply]
when you have this:

```py
new_company = (
   sa.insert(Company)
   .values(name='Acme, Inc.')
   .returning(Company)
   .cte(name='new_company')
)
other_related = (
   sa.select(sa.text('new_company.id'))
   .select_from(new_company)
   .cte('other_related')
)

``` 

you should use the column objects from the first query, like this:

```py
new_company = (
   sa.insert(Company)
   .values(name='Acme, Inc.')
   .returning(Company)
   .cte(name='new_company')
)
other_related = (
   sa.select(new_company.c.id)
   .select_from(new_company)
   .cte('other_related')
)
```

also for things like `text("select * from table")`, that's not immediately compatible with the ORM.  in this case you probably want to say:

```py

fetch_company = (
   sa.select(new_company.select())
   .add_cte(other_related)
)

objects = session.scalars(select(Company).from_statement(fetch_company))
```

aliased() can do this also but since you are working up a full DML type of statement you can use the form shown at https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#using-postgresql-on-conflict-with-returning-to-return-upserted-orm-objects

**Comment by al-dpopowich** [comment]
I have a related, follow up question: if I have a SQL function that returns a record, how can I write a query to return an ORM instance?

With the following SQL (PostgreSQL):

```sql
create table mycomp (
   id serial primary key,
   name varchar not null,
   other varchar,
   stuff varchar
);

create or replace function gen_company()
returns mycomp
as $$
DECLARE
   retval mycomp;
BEGIN
   INSERT INTO mycomp (name)
        SELECT md5(random()::text)
     RETURNING * INTO retval;

   RETURN retval;
END
$$ LANGUAGE plpgsql;

select * from gen_company();
select * from gen_company();
select * from gen_company();
```

With this python:

```python
import sqlalchemy as sa
from sqlalchemy.orm import declarative_base

Base = declarative_base()

_MyComp = sa.Table('mycomp', Base.metadata,
                   sa.Column('id', sa.Integer, primary_key=True),
                   sa.Column('name', sa.String, nullable=False),
                   sa.Column('other', sa.String),
                   sa.Column('stuff', sa.String))

class MyComp(Base):
   __table__ = _MyComp
```


This query strikes me as the most intuitive, select `MyComp` and specify the from-clause, but it generates a cross-join and returns multiple rows when one is expected:

```console
>>> query = sa.select(MyComp).select_from(sa.func.gen_company())
>>> print(query)
SELECT mycomp.id, mycomp.name, mycomp.other, mycomp.stuff
FROM gen_company(), mycomp
```

The following avoids the cross-join, but returns a row of values, not a
single instance of MyComp:

```console
>>> fn = sa.func.gen_company().table_valued(*_MyComp.c)
>>> print(sa.select(fn))
SELECT anon_1.id, anon_1.name, anon_1.other, anon_1.stuff
FROM gen_company() AS anon_1
```

I've made various attempts at using `aliased()`, but can't seem to grok the requirements there, I either end up with a cross-join or an error in evaluating the query.

For the life of me, I can't figure out what seems like something that should be easy: flag a function as returning a row of a table and allow the ORM to use that in determining a proper from clause.

**Reply by zzzeek** [reply]
you might want to try [table_valued](https://docs.sqlalchemy.org/en/20/tutorial/data_select.html#table-valued-functions) for `func.gen_company()` and I would use either `aliased()` or `from_statement`.   you can also try loading the ORM row from `text()` that contains whatever string SQL statement you want, see https://docs.sqlalchemy.org/en/20/orm/queryguide/select.html#getting-orm-results-from-textual-statements

**Reply by al-dpopowich** [reply]
Thanks.  

This seems to do it, though I find it obfuscated:

```console
>>> from_ = sa.select(sa.func.gen_company().table_valued(*_MyComp.c))
>>> query = sa.select(MyComp).from_statement(from_)
>>> print(query)
SELECT anon_1.id, anon_1.name, anon_1.other, anon_1.stuff 
FROM gen_company() AS anon_1
```

**Reply by CaselIT** [reply]
@zzzeek should `sa.select(...).select_from(sa.func.some_function())` raise a deprecation warning?

What's the use case for allowing a function in a from, if table valued is not used?

**Reply by zzzeek** [reply]
that's one way.  here are the two main ways this works right now

```py
from sqlalchemy import Column
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.orm import aliased
from sqlalchemy.orm import declarative_base
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)


a1 = aliased(A, func.my_function().table_valued(*A.__table__.c))

print(select(a1))

print(
    select(A).from_statement(select(func.my_function().table_valued(*A.__table__.c)))
)



```

**Reply by al-dpopowich** [reply]
> a1 = aliased(A, func.my_function().table_valued(*A.__table__.c))

I swear I tried that, but I guess I didn't.  Feeling stupid that I couldn't get that syntax down.  In my defense, it's not immediately obvious what the second argument to `aliased()` should be.  I think I had wrapped the `TableValuedAlias` in a `select()`?  Nor that one can use an `AliasedClass` directly as an argument to `select()`, though I see it now in some examples, but the examples are demoing other functionality, e.g., self-joins.

For the slow among us, it would be wonderful to have the above as an example in the documentation to explicitly demonstrate an arbitrary FROM-clause with a `table_valued()` function.

Thanks again.

**Reply by zzzeek** [reply]
This is a 100% new use case and table_valued() is mostly a PG-only thing.   I was not myself sure that the above examples would even work.    I would say try to work with these two syntaxes and see if they actually scale to being real-world useable for awhile at least.

**Reply by CaselIT** [reply]
@al-dpopowich if you confirm they work on a real word use case a PR that improves the docs to include an example of this use case would be appreciated

**Reply by al-dpopowich** [reply]
Hadn't realized I was being so cutting edge!  :wink: 

This is working in my tests...

I have a SQL function, `clone_company()`, given the `id` of an existing company, and a new name for the company, will copy the company and its related entities.  It returns the row of the newly created company.

```python
async def clone(self, new_name):
      ...
      # alias Company => table-valued function, clone_company()
      clone_fn = func.clone_company(self.id, new_name).table_valued(*self.__table__.c)
      query = select(aliased(Company, clone_fn))

      company = (await self.execute(query)).scalar()

      return company
```

After I get this through QA and into production, I'll look into getting a useful update to the docs and issue a PR.

Thanks.

**Reply by CaselIT** [reply]
That would be great!
